autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,381 @@
+            self.performance_tracker[metric] = {}
+
+    def add_performance(self,
+                        epoch: int,
+                        start_time: float,
+                        end_time: float,
+                        train_loss: float,
+                        train_metrics: Dict[str, float],
+                        val_metrics: Dict[str, float] = {},
+                        test_metrics: Dict[str, float] = {},
+                        val_loss: Optional[float] = None,
+                        test_loss: Optional[float] = None,
+                        ) -> None:
+        """
+        Tracks performance information about the run, useful for
+        plotting individual runs
+        """
+        self.performance_tracker['train_loss'][epoch] = train_loss
+        self.performance_tracker['val_loss'][epoch] = val_loss
+        self.performance_tracker['test_loss'][epoch] = test_loss
+        self.performance_tracker['start_time'][epoch] = start_time
+        self.performance_tracker['end_time'][epoch] = end_time
+        self.performance_tracker['train_metrics'][epoch] = train_metrics
+        self.performance_tracker['val_metrics'][epoch] = val_metrics
+        self.performance_tracker['test_metrics'][epoch] = test_metrics
+
+    def get_best_epoch(self, loss_type: str = 'val_loss') -> int:
+        return np.argmin(
+            [self.performance_tracker[loss_type][e] for e in range(1, len(
+                self.performance_tracker[loss_type]) + 1
+            )]
+        )
+
+    def get_last_epoch(self) -> int:
+        if 'train_loss' not in self.performance_tracker:
+            return 0
+        else:
+            return max(self.performance_tracker['train_loss'].keys())
+
+    def repr_last_epoch(self) -> str:
+        """
+        For debug purposes, returns a nice representation of last epoch
+        performance
+
+        Returns:
+            str: A nice representation of the last epoch
+        """
+        last_epoch = len(self.performance_tracker['train_loss'])
+        string = "\n"
+        string += '=' * 40
+        string += f"\n\t\tEpoch {last_epoch}\n"
+        string += '=' * 40
+        string += "\n"
+        for key, value in sorted(self.performance_tracker.items()):
+            if isinstance(value[last_epoch], dict):
+                # Several metrics can be passed
+                string += "\t{}:\n".format(
+                    key,
+                )
+                for sub_key, sub_value in sorted(value[last_epoch].items()):
+                    string += "\t\t{}: {}\n".format(
+                        sub_key,
+                        sub_value,
+                    )
+            else:
+                string += "\t{}: {}\n".format(
+                    key,
+                    value[last_epoch],
+                )
+        string += '=' * 40
+        return string
+
+
+class BaseTrainerComponent(autoPyTorchTrainingComponent):
+
```

Comment:
A short docstring might be useful here to state that Trainers have to inherit from this class and that `TrainerChoice` will build on top of this

Reply:
I will add this in the next push!